model/tasks/ChangePermissionsTask.php

```diff
@@ -43,32 +44,101 @@
+
         try {
-            $service = $this->getPermissionService();
-            $service->savePermissions(
-                (bool)$params[self::PARAM_RECURSIVE],
+            return $this->doHandle(
                 $this->getClass($params[self::PARAM_RESOURCE]),
-                $params[self::PARAM_PRIVILEGES]
+                (array) $params[self::PARAM_PRIVILEGES],
+                (bool) ($params[self::PARAM_RECURSIVE] ?? false),
+                (bool) ($params[self::PARAM_RECURSIVE_CLASS] ?? false)
             );
-            $result = Report::createSuccess('Permissions saved');
-        } catch (Exception $exception) {
-            $errMessage = sprintf('Saving permissions failed: %s', $exception->getMessage());
+        } catch (Exception $e) {
+            $errMessage = sprintf('Saving permissions failed: %s', $e->getMessage());
+
             $this->getLogger()->error($errMessage);
-            $result = Report::createFailure($errMessage);
+            return Report::createFailure($errMessage);
+        }
+    }
+
+    private function doHandle(
+        core_kernel_classes_Class $root,
+        array $privileges,
+        bool $isRecursive,
+        bool $isRecursiveClass
+    ): Report {
+        if ($isRecursiveClass) {
+            $message = sprintf(
+                "Permissions saved for resources under subclass %s [%s]",
+                $root->getUri(),
+                $root->getLabel()
+            );
+
+            $command = new ChangePermissionsCommand($root, $privileges);
+            $command->withNestedResources();
+
+            $this->getPermissionService()->applyPermissions($command);
+        } elseif ($isRecursive) {
+            $message = 'Starting recursive permissions update';
+
+            $this->createSubtasksForClasses(
+                array_merge(
+                    [$root],
+                    $root->getSubClasses(true) // recursive, NOT including the root
+                ),
+                $privileges
+            );
+        } else {
+            $message = 'Permissions saved';
+
+            $this->getPermissionService()->applyPermissions(
+                new ChangePermissionsCommand($root, $privileges)
+            );
+        }
+
+        return Report::createSuccess($message);
+    }
+
+    private function createSubtasksForClasses(
+        array $allClasses,
+        array $privileges
+    ): void {
+        foreach ($allClasses as $oneClass) {
+            $this->getDispatcher()->createTask(
+                new self(),
+                [
+                    self::PARAM_RESOURCE => $oneClass->getUri(),
+                    self::PARAM_PRIVILEGES => $privileges,
+                    self::PARAM_RECURSIVE => false,
+                    self::PARAM_RECURSIVE_CLASS => true,
+                ],
+                sprintf(
+                    'Processing permissions for class %s [%s]',
+                    $oneClass->getLabel(),
+                    $oneClass->getUri()
```

Comment:
Here we need to make sure we are not creating a string bigger than what DB field supports. Notice that the `label` can big really big (we had issues with this in the past). Better to substr the label or the message, maybe?

Reply:
You are right, the task label in the task_log table is limited to 255, I'll change this :eyes:

Reply:
Done: dae55176b51d927deb924c4e56e154a870a9a462